external-crates/move/move-symbol-pool/src/symbol.rs

```diff
@@ -142,17 +249,121 @@ mod tests {
+        assert_eq!(s2.tag(), Tag::Inlined);
+        assert_ne!(s1.0, s2.0);
+
+        let s1 = Symbol::from("hi");
+        let s2 = Symbol::from("bonjour le monde");
+        assert_eq!(s1.tag(), Tag::Inlined);
+        assert_eq!(s2.tag(), Tag::Dynamic);
+        assert_ne!(s1.0, s2.0);
+
+        let s1 = Symbol::from("bonjour!");
+        let s2 = Symbol::from("bonjour le monde");
+        assert_eq!(s1.tag(), Tag::Dynamic);
+        assert_eq!(s2.tag(), Tag::Dynamic);
         assert_ne!(s1.0, s2.0);
     }
 
     #[test]
     fn test_from_identical_strings_have_the_same_address() {
-        let s1 = Symbol::from("bonjour");
-        let s2 = Symbol::from("bonjour");
+        let s1 = Symbol::from("hi");
+        let s2 = Symbol::from("hi");
+        assert_eq!(s1.tag(), Tag::Inlined);
+        assert_eq!(s2.tag(), Tag::Inlined);
+        assert_eq!(s1.0, s2.0);
+
+        let s1 = Symbol::from("bonjour le monde");
+        let s2 = Symbol::from("bonjour le monde");
+        assert_eq!(s1.tag(), Tag::Dynamic);
+        assert_eq!(s2.tag(), Tag::Dynamic);
         assert_eq!(s1.0, s2.0);
     }
+
+    #[test]
+    fn test_all_static_strings() {
+        // check all symbols round trip correctly
+        for s in STATIC_SYMBOLS {
+            let from_s = Symbol::from(*s);
+            let idx = *STATIC_SYMBOL_IDX.get(*s).unwrap();
+            let from_idx = Symbol::pack_static(idx);
+            assert_eq!(from_s.tag(), Tag::Static);
+            assert_eq!(from_idx.tag(), Tag::Static);
+            assert_eq!(from_s.0, from_idx.0);
+            assert_eq!(from_s.0, from_idx.0);
+            assert_eq!(from_s, from_idx);
+        }
+        // check all entries in `STATIC_SYMBOL_IDX` are present in `STATIC_SYMBOLS`
+        let mut entries = STATIC_SYMBOL_IDX.entries().collect::<Vec<_>>();
+        entries.sort_by_key(|(_, idx)| *idx);
+        assert_eq!(entries.len(), STATIC_SYMBOLS.len());
+        for ((map_s, map_idx), (sym_idx, sym_s)) in
+            entries.iter().zip(STATIC_SYMBOLS.iter().enumerate())
+        {
+            let map_s = *map_s;
+            let map_idx = **map_idx as usize;
+            assert_eq!(map_idx, sym_idx);
+            assert_eq!(map_s, sym_s);
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! check_idx {
+    (@step $_rec:expr,) => {};
+
+    (@step $rec:expr, $head:expr, $str:expr, $($tail:expr,)*) => {
+        assert!(
+            $rec == $head,
+            std::concat!(
+                "Incorrect symbol index for ",
+                $str
+            )
+        );
+        check_idx!(@step $rec + 1usize, $($tail,)*);
+    };
+
+    ($($idx:expr,$str:expr),*) => {
+        check_idx!(@step 0usize, $($idx, $str,)*);
+    }
+}
+
+#[macro_export]
+macro_rules! static_symbols {
+    ($($idx:tt: $str:tt),* $(,)?) => {
+        pub const STATIC_SYMBOLS: &[&str] = &[$($str,)*];
+        pub const STATIC_SYMBOL_IDX: phf::Map<&'static str, u32> = phf::phf_map! {
```

Comment:
Ideally, I wouldn't need 2 pools, but I couldn't figure out how to really access the internals of the perfect hash map. We could just build our own phf data structure maybe? But since I expect the static symbol pool to be "relatively" small, I don't think the 2x allocation here is a big deal